sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -114,73 +71,163 @@ fn verify_length_packable<const VERIFY: bool>(map: &MetadataBTreeMapPrefix) -> R
+    pub fn finish(self) -> Result<MetadataFeature, Error> {
+        let res = MetadataFeature(
+            MetadataBTreeMapPrefix::try_from(
+                self.0
+                    .iter()
+                    .map(|(k, v)| {
+                        Ok((
+                            BoxedSlicePrefix::<u8, MetadataFeatureKeyLength>::try_from(
+                                k.as_bytes().to_vec().into_boxed_slice(),
+                            )
+                            .map_err(|e| Error::InvalidMetadataFeature(e.to_string()))?,
+                            BoxedSlicePrefix::<u8, MetadataFeatureValueLength>::try_from(v.clone().into_boxed_slice())
+                                .map_err(|e| Error::InvalidMetadataFeature(e.to_string()))?,
+                        ))
+                    })
+                    .collect::<Result<MetadataBTreeMap, Error>>()?,
+            )
+            .map_err(Error::InvalidMetadataFeatureEntryCount)?,
+        );
+        verify_packable::<true>(&res)?;
+        Ok(res)
+    }
+}
+
+impl core::ops::Deref for MetadataFeatureMap {
+    type Target = BTreeMap<String, Vec<u8>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl core::ops::DerefMut for MetadataFeatureMap {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
     }
 }
 
 impl StorageScore for MetadataFeature {}
 
 impl WorkScore for MetadataFeature {}
 
-impl TryFrom<Vec<(Vec<u8>, Vec<u8>)>> for MetadataFeature {
-    type Error = Error;
+impl Packable for MetadataFeature {
+    type UnpackError = Error;
+    type UnpackVisitor = ();
 
-    fn try_from(data: Vec<(Vec<u8>, Vec<u8>)>) -> Result<Self, Error> {
-        metadata_feature_from_iter(data)
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        self.0.pack(packer)
+    }
+
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+        visitor: &Self::UnpackVisitor,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        let mut unpacker = CounterUnpacker::new(unpacker);
+        let res = MetadataFeature(
+            MetadataBTreeMapPrefix::unpack::<_, VERIFY>(&mut unpacker, visitor)
+                .map_packable_err(|e| Error::InvalidMetadataFeature(e.to_string()))?,
+        );
+
+        verify_keys_packable::<VERIFY>(&res).map_err(UnpackError::Packable)?;
+        verify_byte_length_packable::<VERIFY>(unpacker.counter()).map_err(UnpackError::Packable)?;
+
+        Ok(res)
     }
 }
 
-impl TryFrom<BTreeMap<Vec<u8>, Vec<u8>>> for MetadataFeature {
+impl TryFrom<Vec<(String, Vec<u8>)>> for MetadataFeature {
     type Error = Error;
 
-    fn try_from(data: BTreeMap<Vec<u8>, Vec<u8>>) -> Result<Self, Error> {
-        metadata_feature_from_iter(data)
+    fn try_from(data: Vec<(String, Vec<u8>)>) -> Result<Self, Error> {
+        let mut builder = Self::build();
+        for (k, v) in data {
+            builder.insert(k, v);
+        }
+        builder.finish()
```

Comment:
Worth it to move this into a function? It's three times the same thing

Reply:
could be, though I don't care too much